text/0077-in-memory-pessimistic-locks.md

```diff
@@ -0,0 +1,232 @@
+
+With this policy, we need to check the `PeerPessimisticLocks` first. If `valid` is `false`, the feature is not usable probably because lock migration is in progress. And if the `epoch` is different from the `epoch` before processing the command, the region must have changed and the lock table we fetched before processing is not the one we should write to. For simplicity, we can just return an `EpochNotMatch` error and let the client retry when any check fails. `epoch` will be increased and `valid` will become `true` when a peer becomes a leader. Later sections will talk more about how `epoch` and `valid ` are maintained.
+
+If the check passes, it is safe for us to write the pessimistic locks into the lock table. After this, we can return the response to the client without proposing anything at all.
+
+For all writes involving the lock CF, the lock in the lock table should be cleared. For example, when a `Prewrite` command replaces a pessimistic lock with a 2PC lock, of course, this is a replicated write, we need to remove the pessimistic lock from the lock table after the write succeeds. This should be done when the raftstore applies any lock CF mutations no matter it is a leader or a follower. For follower, this will not increase much cost because the map should be empty in most of the time.
+
+### Region leader transfer
+
+If it is a voluntary leader transfer triggered by PD, we have the chance to transfer the pessimistic locks to the new leader to avoid unexpected transaction failures.
+
+In `pre_transfer_leader`, the leader serializes current pessimistic locks into bytes and modifies `valid` to `false`. Then, later `AcquirePessimisticLock` commands will fall back to proposing locks. In this way, we can guarantee the pessimistic locks either exist in the serialized bytes, or are replicated through Raft.
+
+The serialized pessimistic locks and the current `propose_index` are carried in the `context` of `MsgTransferLeader` sent by the leader. The context format should be designed to have forward compatibility. When the follower receives the `MsgTransferLeader`, it needs to verify the `propose_index` is smaller than the current `applied_index` before ingesting the locks into its lock table. Otherwise, the follower should reject the leader transfer request.
+
+By default, a Raft message has a size limit of 1 MiB. We will guarantee that the total size of in-memory pessimistic locks in a single region will not exceed the limit. This will be discussed later.
+
+If the transfer leader message is lost or rejected, we need to revert `valid` to `true`. But it is not possible for the leader to know. So, we have to use a timeout to implement it. That means, we can revert `valid` to `true` if the leader is still not transferred after some period. And if the leader receives the `MsgTransferLeader` response from the follower after the timeout, it should ignore the message and not trigger a leader transfer.
```

Comment:
I can not understand. Because raft will only append and apply log in order as they are send, we just need to propose the  entry `MsgTransferLeader` after `lock`. Just like that.

```rust
fn propose_transfer_leader(&mut self, msg: RaftMessage) {
    if !self.locks.is_empty() {
           let lock_msg = RaftMessage::new();
           for lock in self.locks.take() {
                 lock_msg.append(lock);
                 if lock_msg.size() > 1024 * 1024 {   // 1MB
                     self.propose(lock_msg);
                 }
           }
           self.propose(lock_msg);
     }
     self.propose(msg);
}

```

Reply:
Well, my initial though was to not change the transfer leader procedure. Using a proposal is simpler.

Reply:
Another difference is the impact of leader transafer on ongoing pessimistic transactions such as the fallback processing described in this document. If the impact is ok we could use the simpler way and make it a whole, it's easier to maintain.

Reply:
The timeout mechanism is still needed. The leader transfer timeout is election timeout in raft. We can check it on tick.